File: 2296-design-a-text-editor/2296-design-a-text-editor.js

var TextEditor = function() {
  this.left = [];
  this.right = [];
};

/** 
 * @param {string} text
 * @return {void}
 */
TextEditor.prototype.addText = function(text) {
  for (let ch of text) this.left.push(ch);
};

/** 
 * @param {number} k
 * @return {number}
 */
TextEditor.prototype.deleteText = function(k) {
  let cnt = 0;
  while (this.left.length && cnt < k) {
    this.left.pop();
    cnt++;
  }
  
  return cnt;
};

/** 
 * @param {number} k
 * @return {string}
 */
TextEditor.prototype.cursorLeft = function(k) {
  while (this.left.length && k--) 
    this.right.push(this.left.pop());
  
  return this.left.slice(Math.max(this.left.length - 10, 0)).join('');
};

/** 
 * @param {number} k
 * @return {string}
 */
TextEditor.prototype.cursorRight = function(k) {
  while (this.right.length && k--) 
    this.left.push(this.right.pop());
  
  return this.left.slice(Math.max(this.left.length - 10, 0)).join('');
};

/** 
 * Your TextEditor object will be instantiated and called as such:
 * var obj = new TextEditor()
 * obj.addText(text)
 * var param_2 = obj.deleteText(k)
 * var param_3 = obj.cursorLeft(k)
 * var param_4 = obj.cursorRight(k)
 */

File: 2296-design-a-text-editor/2296-design-a-text-editor.py
class TextEditor:

    def __init__(self):
        self.left = []
        self.right = []

    def addText(self, text: str) -> None:
        self.left += list(text)

    def deleteText(self, k: int) -> int:
        cnt = 0
        while self.left and cnt < k:
            self.left.pop()
            cnt += 1
        return cnt

    def cursorLeft(self, k: int) -> str:
        while self.left and k:
            self.right.append(self.left.pop())
            k -= 1
        return ('').join(self.left[max(len(self.left)-10, 0):])

    def cursorRight(self, k: int) -> str:
        while self.right and k:
            self.left.append(self.right.pop())
            k -= 1
        return ('').join(self.left[max(len(self.left)-10, 0):])
        


# Your TextEditor object will be instantiated and called as such:
# obj = TextEditor()
# obj.addText(text)
# param_2 = obj.deleteText(k)
# param_3 = obj.cursorLeft(k)
# param_4 = obj.cursorRight(k)